r/R/dplyr-funcs-type.R

```diff
@@ -76,6 +76,53 @@ register_bindings_type_cast <- function() {
   register_binding("as.numeric", function(x) {
     Expression$create("cast", x, options = cast_options(to_type = float64()))
   })
+  register_binding("as.Date", function(x,
+                                       format = NULL,
+                                       origin = "1970-01-01",
+                                       tz = "UTC") {
+    # base::as.Date() first converts to UTC and then extracts the date, which is
+    # why we need to go through timestamp() first - see unit tests for the real
+    # life impact of the difference between lubridate::date() and base::as.Date()
+    # browser()
+    if (call_binding("is.Date", x)) {
+      # arrow_date <- build_expr("cast", x, options = cast_options(to_type = date32()))
+      return(x)
+    } else if (call_binding("is.POSIXct", x)) {
+      if (tz == "UTC") {
+        arrow_timestamp <- build_expr("cast", x, options = cast_options(to_type = timestamp(timezone = tz)))
+        return(build_expr("cast", arrow_timestamp, options = cast_options(to_type = date32())))
+      } else {
+        abort("`as.Date()` with a timezone different to 'UTC' is not supported in Arrow")
+      }
+    } else if (call_binding("is.character", x)) {
+      # this could be improved with tryFormats once strptime returns NA and we
+      # can use coalesce - https://issues.apache.org/jira/browse/ARROW-15659
+      # TODO revisit once https://issues.apache.org/jira/browse/ARROW-15659 is done
+      if (!is.null(format)) {
+        arrow_timestamp <- call_binding("strptime", x, format, unit = "s")
+        return(build_expr("cast", arrow_timestamp, options = cast_options(to_type = date32())))
+      } else {
```

Comment:
I suspect we actually will need to do something about these as it's causing failures elsewhere: https://github.com/apache/arrow/runs/5335594462?check_suite_focus=true#step:7:25913

We could probably get away with defaulting the format to `%Y-%m-%d` and note that we don't (yet) support `%Y/%m/%d` as a default.

Another option might be to identify if we're working with an R object or an Arrow object, and specifically call `base::as.Date()` if it's an R object. (this approach might actually be even broader and useful elsewhere for cases like this). For single-values this should be fine because converting from an R object of length 1 and then doing an arrow computation is not much faster than doing a computation on an R object of length 1 and then doing the R->Arrow conversion.

Reply:
My plan was to revisit this once the `strptime()` kernel supports returning NAs and then `coalesce()` through the `tryFormats` `base::as.Date()` defaults. In the meantime, defaulting to `"%Y-%m-%d"` sounds like a good plan.

Reply:
Solved. I've added the `tryFormats` arg but with a single option - we're not trying all the supplied formats.